File: sysrap/tests/sblackbody_test.py
#!/usr/bin/env python

import os, numpy as np
from opticks.ana.fold import Fold
import matplotlib.pyplot as mp
SIZE=np.array([1280, 720])
TEST = os.environ.get("TEST","")


def planck_spectral_radiance_set(f):
    psr = f.planck_spectral_radiance_set
    psr_names = f.planck_spectral_radiance__set_names

    print(psr)
    print(psr_names)

    fig, ax = mp.subplots(figsize=SIZE/100.)
    title= f"sblackbody_test.sh : {f.base} : {TEST}  "
    fig.suptitle(title)
    for i in range(len(psr)):
        nm = psr[i,:,0]
        bb = psr[i,:,1]
        bb /= bb.sum()
        ax.plot( nm, bb, label=psr_names[i] )
    pass
    ax.legend()
    fig.show()


def planck_cdf(f):
    cdf = f.cdf
    print(f"planck_cdf cdf.shape {cdf.shape}")
    fig, ax = mp.subplots(figsize=SIZE/100.)
    title= f"sblackbody_test.sh : {f.base} : {TEST}  "
    fig.suptitle(title)
    ax.plot( cdf[:,0], cdf[:,1], label="cdf" )
    ax.legend()
    fig.show()

def planck_icdf_0(f):
    icdf = f.icdf.reshape(-1)
    print(f"planck_icdf_0 icdf.shape {icdf.shape}")
    fig, ax = mp.subplots(figsize=SIZE/100.)
    title= f"sblackbody_test.sh : {f.base} : {TEST}  "
    fig.suptitle(title)

    idom = np.linspace(0,1, len(icdf) )
    ax.plot( idom, icdf, label="icdf" )
    ax.legend()
    fig.show()

def planck_icdf(f):
    icdf_prop = f.icdf_prop
    print(f"planck_icdf icdf_prop.shape {icdf_prop.shape}")
    fig, ax = mp.subplots(figsize=SIZE/100.)
    title= f"sblackbody_test.sh : {f.base} : {TEST}  "
    fig.suptitle(title)

    ax.plot( icdf_prop[:,0], icdf_prop[:,1], label="icdf_prop" )
    ax.legend()
    fig.show()

def planck_sample(f):
    sample = f.wavelength
    print(f"planck_sample sample.shape {sample.shape}")
    fig, ax = mp.subplots(figsize=SIZE/100.)
    title= f"sblackbody_test.sh : {f.base} : {TEST}  "
    fig.suptitle(title)

    w = np.arange(80.,801,.1, dtype=np.float64)
    wbin = w[::100]

    hn, hd = np.histogram(sample, wbin)
    assert np.all(hd == wbin)
    assert len(hn) == len(wbin) - 1   # looses one bin

    ax.plot( hd[:-1], hn , drawstyle="steps-post", label="sample")  # -pre -mid -post

    ax.legend()
    fig.show()



if __name__ == '__main__':
    f = Fold.Load(symbol="f")
    print(repr(f))

    if TEST == "planck_spectral_radiance_set":
        planck_spectral_radiance_set(f)
    elif TEST == "planck_cdf":
        planck_cdf(f)
    elif TEST == "planck_icdf":
        planck_icdf(f)
    elif TEST == "planck_sample":
        planck_sample(f)
    else:
        print(f"TEST {TEST} unhandled")
    pass

File: sysrap/tests/sblackbody_test.cc
#include "ssys.h"
#include "sblackbody.h"
#include "NPFold.h"

struct sblackbody_test
{
    static int planck_spectral_radiance_set();
    static int planck_sample();

    static int main();
};


inline int sblackbody_test::planck_spectral_radiance_set()
{
    int nj = 2000;
    NP* a = sblackbody<double>::make_planck_spectral_radiance_set(nj);
    a->save("$FOLD/planck_spectral_radiance_set.npy");
    return 0 ;
}

inline int sblackbody_test::planck_sample()
{
    sblackbody<double> bb(4096,4096,6500.,80.,800., 1'000'000 );
    bb.save("$FOLD");
    return 0 ;
}

inline int sblackbody_test::main()
{
    const char* TEST = ssys::getenvvar("TEST","ALL");
    bool ALL = 0 == strcmp(TEST, "ALL") ;
    int rc = 0;
    if(ALL||0==strcmp(TEST,"planck_spectral_radiance_set")) rc += planck_spectral_radiance_set();
    if(ALL||0==strcmp(TEST,"planck_sample")) rc += planck_sample();
    return rc ;
}

int main()
{
    return sblackbody_test::main();
}
